### DISCUSSION TITLE
Setup nvm

### DISCUSSION THREAD

**Discussion by fellnerse** [nushell/nushell#4220] [Q&A] [answered]
Hey,

I'm currently migrating to nushell and wondering how to setup different tools. i.e. now I try to use nvm. Previously I had nvm setup in fish with bass (emulation of bash and forwarding the env variables). Sourcing the `nvm.sh` does not work for me, it just does not return.

Does someone have an idea?

Thanks in advance

**Comment by kubouch** [comment, marked as answer]
Hi!

Unfortunately, it seems nvm.sh only targets POSIX shells. Nushell does not support POSIX and does not have any POSIX emulation layer like bass. You can *run* Bash scripts just fine (`bash file.sh` or `bash -c ...commands...`) but it won't capture the environment variables.

I think you might need to write a wrapper around nvm.sh like people did for fish. Some of the fish wrappers seem quite simple, like https://github.com/brigand/fast-nvm-fish or https://github.com/FabioAntunes/fish-nvm so it might not be that hard.

**Reply by fellnerse** [reply]
Would it make sense to have a general wrapper for nu around any bash script like bass?

**Reply by fellnerse** [reply]
and would it be possible to solve this via a plugin?

**Reply by kubouch** [reply]
What we could do in Nushell is to add a way to capture environment variables of a process. This is how bass seems to work and you should be able to do that with the sysinfo crate in Rust as well. So in Nushell, you would call, for example, `env capture bash -c "source file.sh"` which would return you a list of environment variables of the `bash` process. Then, you could write a wrapper that would apply these changes on top of the current environment.

The `env capture` could be implemented even now as a plugin. However, I can see it also being a part of core Nushell since people are asking a lot about POSIX compatibility. The wrapper itself then could be just a .nu script or again a plugin.

If you plan working on this, note that we're now rewriting a significant part of Nushell (we call it the engine-q project). Plugins are one of the things that were reworked and will be replaced with the new system once we merge the changes into the main Nushell repository. We also changed a bit how environment variables work and added a module system that might be useful for this use case.

**Reply by hustcer** [reply]
Try `fnm` instead: https://github.com/Schniz/fnm, faster!

**Reply by fellnerse** [reply]
Thank you for both of your answers! I guess something like this `env capture` would be extremely useful for a lot of people and if you need help I could try to give it a go.

Also will try out `fnm`

**Reply by fellnerse** [reply]
@hustcer did you try do install the shell completions? if so what did you do?

**Reply by hustcer** [reply]
Unfortunately, shell completion is not well supportted

**Comment by Thlnking** [comment]
@hustcer @fellnerse 
https://dev.to/vaibhavdn/using-fnm-with-nushell-3kh1 
Maybe this blog article can help your use fnm in unshell.

**Reply by hustcer** [reply]
Great! Thanks

**Reply by DevYatsu** [reply]
If the previous article seems too complex or just doesn't work, just check [this comment](https://github.com/Schniz/fnm/issues/463#issuecomment-2987700729), one copy-paste and you're done
